### PR TITLE
Case sensitive changes

### PR DESCRIPTION
**Pull request by medianotion**
Changed the code to do a case sensitive compare and allow 0 - 255 ASCII characters.

### CODE REVIEW

File: Permutation.py [hunk 1]
```diff
@@ -1,33 +1,29 @@
...
-    # modify 1 copy and you modify them all.
-    # considered using a dictionary and using the ascii code as the key and a 2 dim list as the value.  
-    # but I thought the dictionary search for the key to increment the character count would be a search on the dictionary and therefore n log n
-    charCountMap = [] 
-    for index in range(256):
-        charCountMap.append([0,0])
-
-    if (IsValidateString(str1) & IsValidateString(str2)):
+    if (len(str1) != len(str2)):
+        return False
```
Comment (medianotion) [inline Permutation.py]: Removed the stripping code.  Changed the CharCountMap to a single dim list/array

File: Permutation.py [hunk 2]
```diff
@@ -37,19 +33,19 @@ def IsValidateString(stringToCheck):
...
+def CountCharactersInStrings(inputString1, inputString2, charCountMap):
 
-def main():
-    str1 = "!@$"
-    str2 = "$@!"
+    for index in range(len(inputString1)):
+        asciiCode1 = ord(inputString1[index])
+        asciiCode2 = ord(inputString2[index])
+        charCountMap[asciiCode1] = charCountMap[asciiCode1] + 1
+        charCountMap[asciiCode2] = charCountMap[asciiCode2] - 1
```
Comment (medianotion) [inline Permutation.py]: Updated the CountCharactersInStrings method to increment the list value for inputString1 and decrement the list value for inputString2.

**Comment by medianotion** [comment]
@jrpdrummer please take a look at the latest branch code.  I added unit tests, made the suggested changes to check string lengths and to use the single dim list and increment/decrement values.  Finally, looped over the list and if everything is 0 it is a permutation.  Otherwise it is not.

File: Permutation.py [hunk 3]
```diff
@@ -1,33 +1,29 @@
...
-    # [[0,0]] * 256 makes each element a copy of the first.  
-    # modify 1 copy and you modify them all.
-    # considered using a dictionary and using the ascii code as the key and a 2 dim list as the value.  
-    # but I thought the dictionary search for the key to increment the character count would be a search on the dictionary and therefore n log n
-    charCountMap = [] 
-    for index in range(256):
-        charCountMap.append([0,0])
-
-    if (IsValidateString(str1) & IsValidateString(str2)):
+    if (len(str1) != len(str2)):
```
Comment (jrpdrummer) [inline Permutation.py]: You probably want to do this after your None/Empty string check.

File: Permutation.py [hunk 4]
```diff
@@ -1,33 +1,29 @@
...
+        return False
 
-        CountCharactersInString(str1,charCountMap,0)
-        CountCharactersInString(str2,charCountMap,1)
+    if (IsValidateString(str1) and IsValidateString(str2)):
```
Comment (jrpdrummer) [inline Permutation.py]: You can reduce nesting by inverting the logic here (e.g. "if not valid(s1) or not valid(s2)"). Not necessary, but can make code cleaner.

File: Permutation.py [hunk 5]
```diff
@@ -1,33 +1,29 @@
...
 
+        # init a list with 256 elements set to 0
+        # for the life of me I couldn't figure out how to initialize this list without a loop.
```
Comment (jrpdrummer) [inline Permutation.py]: Same initialization using a list comprehension:
`charCountMap = [0 for x in range(0,256)]`

File: test.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+    def test_not_equal_length_fail(self):
+        self.assertFalse(Permutation.ArePermutations('ABC', 'AB'))
+
+    def test_empty_1_string_fail(self):
+        self.assertFalse(Permutation.ArePermutations('ABC', ''))
+
+    def test_empty_2_string_fail(self):
+        self.assertFalse(Permutation.ArePermutations('', ''))
+
```
Comment (jrpdrummer) [inline test.py]: Missing a test for Permutation.ArePermutations(None, None).

**Comment by jrpdrummer** [comment]
I have left a final set of comments. Please review.

File: Permutation.py (hunk 3 above)
Comment (medianotion) [reply Permutation.py]: Why?  Is the len more expensive than a None/Empty check?

File: Permutation.py (hunk 4 above)
Comment (medianotion) [reply Permutation.py]: I can see why that will simplify the if/else

File: test.py (hunk 6 above)
Comment (medianotion) [reply test.py]: Dooh.  OK.

File: Permutation.py (hunk 3 above)
Comment (jrpdrummer) [reply Permutation.py]: len doesn't work on None:
`TypeError: object of type 'NoneType' has no len()`

**Pull request merged**